miner/src/miner.rs

```diff
@@ -82,26 +82,26 @@ impl Miner {
     }
 
     /// TODO(doc): @quake
-    // remove `allow` tag when https://github.com/crossbeam-rs/crossbeam/issues/404 is solved
-    #[allow(clippy::zero_ptr, clippy::drop_copy)]
     pub fn run(&mut self) {
         loop {
             select! {
                 recv(self.work_rx) -> msg => match msg {
                     Ok(work) => {
-                        let pow_hash= work.block.header().calc_pow_hash();
-                        let (target, _,) = compact_to_target(work.block.header().raw().compact_target().unpack());
-                        self.works.put(pow_hash.clone(), work);
-                        self.notify_workers(WorkerMessage::NewWork{pow_hash, target});
+                        match work {
+                            Works::FailSubmit(hash) => {
+                                self.legacy_work.pop(&hash);
+                            },
+                            Works::New(work) => self.submit_work(work),
```

Comment:
This fn name is very misleading, suggest to rename to `notify_new_work`
```suggestion
                            Works::New(work) => self.notify_new_work(work),
```